pkg/networkservice/common/mechanisms/vxlan/vni/server.go

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2020-2021 Cisco and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vni
+
+import (
+	"context"
+	"net"
+	"sync"
+
+	"github.com/golang/protobuf/ptypes/empty"
+
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"github.com/networkservicemesh/api/pkg/api/networkservice/mechanisms/vxlan"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+)
+
+type vniServer struct {
+	tunnelIP net.IP
+	sync.Map
+}
+
+// NewServer - set the DstIP *and* VNI for the vxlan mechanism
+func NewServer(tunnelIP net.IP) networkservice.NetworkServiceServer {
+	return &vniServer{
+		tunnelIP: tunnelIP,
+	}
+}
+
+func (v *vniServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
+	if mechanism := vxlan.ToMechanism(request.GetConnection().GetMechanism()); mechanism != nil {
+		mechanism.SetDstIP(v.tunnelIP)
+		k := vniKey{
+			srcIPString: mechanism.SrcIP().String(),
+			vni:         mechanism.VNI(),
+		}
+		// If we already have a VNI, make sure we remember it, and go on
+		if k.vni != 0 && mechanism.SrcIP() != nil {
+			v.Map.LoadOrStore(k, &k)
+			return next.Server(ctx).Request(ctx, request)
```

Comment:
Probably we need to add here check for returned error and `v.Map.Delete()` if so.

Reply:
yes! done.

Reply:
So actually I have meant something more specific:
```go
_, loaded := v.Map.LoadOrStore(k, &k)
conn, err := next.Server(ctx).Request(ctx, request)
if err != nil {
    if !loaded {
        v.Map.Delete(k)
    }
    return nil, err
}
```
If it was the 1 request for the Connection, it wouldn't be healed or closed on timeout, so we need this `Delete()`. But for the refresh case we probably may want to keep it until the Connection will be restored with heal or closed by heal/timeout.

Reply:
done.